### ISSUE TITLE
issue with `q` and parsing in PGML

### ISSUE THREAD

**Issue by Alex-Jordan** [openwebwork/pg#951]
This is a MWE that I boiled down from a problem that is throwing errors now. (If that problem threw errors in the past, we never saw them. It could have been the kind of error that went to the apache error log in the past.)

```
DOCUMENT();
loadMacros(qw(PGstandard.pl PGML.pl));

BEGIN_PGML
[@ %f = (q => 0) @]
END_PGML
ENDDOCUMENT();
```

If you change the `q` to something else, it seems to always work as expected. So I suspect this has to do with the special parsing for things like `q(string)`.

This MWE is silly, but the real example had something where a function was passed something like `q => 0` as an option. I can easily change the problem code to get around the errors. But I'm wondering if this is a bug that can be fixed or if it is an unavoidable consequence of the special parsing for `q` constructions.

**Comment by drgrice1** [comment]
If you change the definition of `$quoted` on line 46 of PGML.pl to be the empty string (and thus disabling the new `q(string)` parsing), then your MWE works without problem.  I think that is a pretty definitive test that this is the new parsing.  Of course this doesn't help much, since you were probably already convinced of this.

**Comment by drgrice1** [comment]
Okay, a better is test is to remove `$quoted` from the `$splitPattern`.  Changing `$quoted` to the empty string was not right.

In any case, the specific part of `$quoted` that is causing the issue is `\bq[qr]?\s+(?=.)`.

**Comment by dpvc** [comment]
Try changing the pattern to

``` perl
my $quoted    = '[$@%]q[qr]?|\bq[qr]?\s+(?|=>)(?=.)|\bq[qr]?(?|=>)(?=\W)';
```

(with two additions of `(?|=>)`) and see if that does the trick (without disrupting anything else).

**Comment by dpvc** [comment]
OOPS, I meant `(?!=>)` not `(?|=>)`.  I have fixed it above.

**Comment by Alex-Jordan** [comment]
This fixes the issue, thanks. Now I could have the following line in the PGML:

```
[@ %f = (q => 0) @]
```

It got me thinking though, about where else it is legal to have a bare `q`. This led to trying to follow the above with:

```
[@ $f{q} @]
```

and it creates the same error message. Ideally we wouldn't do the same thing with the `}` as you did with `=>` because a legal string constructor is `q}string}`. I have some thoughts about what could be done but it starts to feel like whack-a-mole.

**Comment by drgrice1** [comment]
I was just testing this and did the same thing you did, and got the same error.

It is interesting that `q}string}` is a legal construct.  According the perl documentation "non-bracketing delimiters use the same character fore and aft, but the four sorts of ASCII brackets (round, angle, square, curly) all nest".  Of course this does not say you can't use an ending delimiter to start the quote, and according to that statement the same character should end the quote.  In testing I see that it does work in a perl script.

**Comment by Alex-Jordan** [comment]
Of course, no reasonable person would code `q}string}`. With some PTX code where the `string` is automated, we look through `string` and identify an unused character to be a delimiter sometimes. It has to be an unused character from a prescribed list of characters. And `}` is not on that list.

I guess I'm saying that if there's nothing better that comes to light, I'm fine with handling `}` just like Davide handled `=>`.

In researching this, I took a bit of a dive into these constructions. Another little thing I saw is that `q#string#` is legal and makes a string, but `q #string#` will treat that first `#` as starting a comment. And it takes the first non-whitespace character of the next line as the delimiter. My point is not that PGML should replicate this behavior. But rather that there are some tough nitty-gritty things with these constructions and I'll be fine without making PGML handle them all the same way that perl does. I'm just grateful to have a way to do quotes when `'` or `"` are unavailable.

**Comment by dpvc** [comment]
OK, here is a patch that should take care of the issues Alex has raised.

``` diff
diff --git a/macros/core/PGML.pl b/macros/core/PGML.pl
index db587a47..772d0c85 100644
--- a/macros/core/PGML.pl
+++ b/macros/core/PGML.pl
@@ -43,7 +43,7 @@ my $list      = '(?:^|(?<=[\t ]))(?:[-+o*]|(?:\d+|[ivxl]+|[IVXL]+|[a-zA-Z])[.)])
 my $align     = '>> *| *<<';
 my $code      = '```';
 my $pre       = ':   ';
-my $quoted    = '[$@%]q[qr]?|\bq[qr]?\s+(?=.)|\bq[qr]?(?=\W)';
+my $quoted    = '[$@%]q[qr]?|\bq[qr]?\s+(?:#.*?(?:\n\s*)+)?(?!=>)(?=.)|\bq[qr]?(?!=>)(?=\W)';
 my $emphasis  = '\*+|_+';
 my $chars     = '\\\\.|[{}[\]()\'"]';
 my $ansrule   = '\[(?:_+|[ox^])\]\*?';
@@ -547,9 +547,9 @@ sub Quoted {
        my $quote = substr($next, 0, 1);
        $self->{split}[ $self->{i} ] = substr($next, 1);
        my $pcount = 0;
-       my $open   = ($quote =~ m/[({[]/ ? $quote : '');
+       my $open   = ($quote =~ m/[({[<]/ ? $quote : '');
        my $close  = $open || $quote;
-       $close =~ tr/({[/)}]/;
+       $close =~ tr/({[</)}]>/;
        my $qclose = "\\$close";
        $self->Text($token . $quote);
 
@@ -701,7 +701,7 @@ my $balanceAll = qr/[\{\[\'\"]/;
                parseQuoted        => 1,
                terminator         => qr/@\]/,
                terminateMethod    => 'terminateGetString',
-               balance            => qr/[\'\"]/,
+               balance            => qr/[\{\'\"]/,
                allowStar          => 1,
                allowDblStar       => 1,
                allowTriStar       => 1
```

The `(?:#.*?(?:\n\s*)+)?` takes care of the `#` as comment.  I also hadn't known that `<` and `>` act as matched delimiters, so I've added those in the two lines that handle that.  Finally, the addition of `\{` to the `balance` pattern handles the issue with `$f{q}`.  I don't think that should cause any other issues, but let me know if you see any other consequences from that last one.

**Comment by Alex-Jordan** [comment]
Thanks again @dpvc , I tested with the following and it works.

```
DOCUMENT();
loadMacros(qw(PGstandard.pl PGML.pl));

BEGIN_PGML
[@ %f = (q => 0) @]

[@ $f{q} @]

[@ q#taco# @]

[@ q #taco#
pizzap @]
END_PGML
ENDDOCUMENT();
```

**Comment by Alex-Jordan** [comment]
This was fixed by #959.
